src/commands/watch.rs

```diff
@@ -145,116 +173,129 @@ impl WatchCommand {
+        }
+        spin_args
     }
 
-    async fn generate_path_patterns(&self) -> Result<Vec<String>> {
-        // TODO: Spend some time cleaning up this function
+    async fn generate_filter_config(&self) -> Result<crate::watch_filter::Config> {
+        let app_dir = parent_dir(&self.app)?;
+        let builder = WatchPatternBuilder::new(&app_dir);
         let app_manifest = spin_loader::local::raw_manifest_from_file(&self.app)
             .await?
             .into_v1();
 
-        let watch_patterns = app_manifest
+        // We always want to watch the application manifest
+        let manifest_pattern = builder.watched(
+            self.app
+                .to_str()
+                .with_context(|| format!("non-unicode manifest path {:?}", self.app))?
+                .to_owned(),
+        )?;
+
+        // We want to watch the source code if we aren't skipping the build step
+        let source_patterns = app_manifest
             .components
             .iter()
             .filter_map(|c| {
-                if let Some(build) = c.build.as_ref() {
-                    match build.watch.clone() {
-                        Some(watch) => {
-                            let Some(workdir) = build.workdir.clone() else {
-                                return Some(watch);
-                            };
-                            Some(
-                                watch
-                                    .iter()
-                                    .filter_map(|ws| workdir.join(ws).to_str().map(String::from))
-                                    .collect::<Vec<String>>(),
-                            )
-                        }
-                        None => {
-                            eprintln!(
-                                "You haven't configured what to watch for the component: '{}'. Learn how to configure Spin watch at https://developer.fermyon.com/common/cli-reference#watch",
-                                c.id
-                            );
-                            None
-                        }
-                    }
-                } else {
-                    // No build config for this component so lets watch the source instead
-                    if let RawModuleSource::FileReference(path) = &c.source {
-                        if let Some(string_path) = path.to_str() {
-                            return Some(vec![String::from(string_path)]);
-                        }
-                    }
-                    None
-                }
+                let Some(build) = c.build.as_ref() else {
+                    return None;
+                };
+                let Some(watch) = build.watch.clone() else {
+                    eprintln!(
+                        "You haven't configured what to watch for the component: '{}'. Learn how to configure Spin watch at https://developer.fermyon.com/common/cli-reference#watch",
+                        c.id
+                    );
```

Comment:
If a component doesn't have a build section then it will never get to this warning. AFAICT it is reasonable to warn with this on any component that has a build section.

Reply:
I guess its a little annoying for my pet use-case of watching just the `source`. If I want to suppress this warning I have to copy `source` into `watch`.

Reply:
Now I understand. I get why that could be annoying for that use case. I'll file an issue.